README.md
# 说明

使用这个库你可以实现类似于[SnackBar](https://flutter.dev/docs/cookbook/design/snackbars)类似的提示，不同的是`flutter_snackbar`的提示显示在顶部。

效果如下：  
![flutter_snackbar](assets/flutter_snackbar.gif)

## 示例

```dart
import 'package:flutter_snackbar/flutter_snackbar.dart';

// 创建SnackBarWidget
SnackBarWidget(
    // 绑定GlobalKey，用于调用显示/隐藏方法
    key: _globalKey,
    // 设置动态变化的Text
    textBuilder: (String message) {
      return Text(message ?? "",
          style: TextStyle(color: Colors.white, fontSize: 16.0));
    },
    // 内容不变时使用text属性
    text: Text("内容不变时使用text属性"),
    // 设定背景decoration
    decoration: ShapeDecoration(
        shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.all(Radius.circular(20))),
        color: Colors.blue.withOpacity(0.8)),
    // 用于显示内容，默认是填充空白区域的
    content: Center(child: Text("这是内容部分")))
```

显示显示SnackBar：

```dart
_globalKey.currentState.show("这是SnackBar count: ${count++}");
```

隐藏SnackBar：

```dart
_globalKey.currentState.dismiss();
```


CHANGELOG.md
## [0.0.1]  
`SnackBar`初步完成。

## [0.0.2]
增加了部分注释。

## [0.0.3]
修改了获取`Size`的方式，防止在`build`过程中无法正常获取到宽高属性。

lib/flutter_snackbar.dart
library flutter_snackbar;

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

/// 用于动态构建[Text]，以实现动态改变[SnackBarWidget]中内容的目的
typedef TextBuilder = Text Function(String message);

/// 用于展示类似于Android中SnackBar的提示框
class SnackBarWidget extends StatefulWidget {
  /// 提示框中要显示的内容，提示内容不需要发生变化时使用
  final Text text;

  /// 用于动态构建Text，当需要动态改变SnackBarWidget的内容时使用
  final TextBuilder textBuilder;

  /// 主题内容部分
  final Widget content;

  /// 提示框的内边距
  final EdgeInsets padding;

  /// 提示框的外边距
  final EdgeInsets margin;

  /// 提示框持续时间（从淡入到淡出）
  final Duration duration;

  /// 提示框的包装样式
  final Decoration decoration;

  SnackBarWidget(
      {Key key,
      this.text,
      this.textBuilder,
      this.content,
      this.padding,
      this.margin,
      this.duration,
      this.decoration})
      : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return SnackBarWidgetState();
  }
}

class SnackBarWidgetState extends State<SnackBarWidget>
    with TickerProviderStateMixin {
  final GlobalKey _snackKey = GlobalKey();
  GlobalKey<_DynamicTextState> _textKey;

  AnimationController _controller;

  bool get isShowing => _controller.status != AnimationStatus.dismissed;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
        duration: widget.duration ?? Duration(milliseconds: 1400), vsync: this);
  }

  @override
  void dispose() {
    // 此时进行资源回收
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      alignment: AlignmentDirectional.topCenter, // 顶部居中
      fit: StackFit.loose, // 如果child没有指定位置，则采用使用child自身的大小
      children: <Widget>[
        // 使内容部分填充剩余空间
        SizedBox.expand(child: widget.content),
        SnackBarAnimation(
            key: _snackKey,
            controller: _controller,
            child: Container(
              // 如果存在textBuilder，则创建_DynamicText，用来动态更新Text的内容
              // 如果不存在textBuilder，则使用静态的Text
              child: widget.textBuilder != null
                  ? _DynamicText(
                      widget.textBuilder,
                      key: _textKey = GlobalKey(),
                    )
                  : widget.text,
              padding:
                  widget.padding ?? EdgeInsets.fromLTRB(15.0, 10.0, 20.0, 10.0),
              margin: widget.padding ?? EdgeInsets.only(top: 10.0),
              decoration: widget.decoration ??
                  ShapeDecoration(
                      shape: RoundedRectangleBorder(
                          borderRadius:
                              BorderRadius.all(Radius.circular(30.0))),
                      color: Colors.orange),
            ))
      ],
    );
  }

  /// 显示SnackBar
  /// [message] 要更新的提示内容
  void show([String message]) {
    if (_textKey != null && _textKey.currentState != null) {
      _textKey.currentState.update(message);
    }
    (_snackKey.currentWidget as SnackBarAnimation).playAnimation();
  }

  /// 隐藏SnackBar
  void dismiss() {
    if (_controller.isDismissed) return;
    (_snackKey.currentWidget as SnackBarAnimation).reverseAnimation();
  }
}

class SnackBarAnimation extends StatelessWidget {
  final AnimationController controller;
  final Container child;
  Animation<double> fade;
  Animation<double> translate;

  SnackBarAnimation(
      {@required GlobalKey key, @required this.controller, this.child})
      : assert(key != null), // 由于需要通过BuildContext来获取Widget的高度，此处的key为必须的
        assert(controller != null),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      builder: _buildAnimation,
      animation: controller,
      child: child,
    );
  }

  // 开始播放动画
  Future<Null> playAnimation() async {
    // 此处通过key去获取Widget的Size属性
    RenderBox renderBox = (key as GlobalKey).currentContext.findRenderObject();
    Size size = renderBox.size;
    double deltaY = size.height; // 该值为位移动画需要的位移值

    // 如果fade动画不存在，则创建一个新的fade动画
    fade = fade ??
        Tween<double>(begin: 0.0, end: 1.0).animate(CurvedAnimation(
            parent: controller,
            curve: Interval(0.0, 0.3, // 持续时间为总持续时间的30%
                curve: Curves.ease)));

    translate = translate ??
        Tween<double>(begin: -deltaY, end: 0).animate(CurvedAnimation(
            parent: controller,
            curve: Interval(0.0, 0.15, curve: Curves.ease))); // 前15%的时间用于执行平移动画

    try {
      await controller.forward().orCancel;
      await controller.reverse().orCancel;
    } on TickerCanceled {}
  }

  Future<Null> reverseAnimation() async {
    try {
      await controller.reverse().orCancel;
    } on TickerCanceled {}
  }

  Widget _buildAnimation(BuildContext context, Widget child) {
    return Transform.translate(
      child: Opacity(
          child: child,
          opacity: fade != null
              ? fade.value
              : 0), // 此处使用fade.value不断取值来刷新child的opacity
      offset: Offset(
          0,
          translate != null
              ? translate.value
              : 0), // 此处使用translate.value不断取值来刷新child的偏移量
    );
  }
}

/// 能够动态更新内容的[Text]
class _DynamicText extends StatefulWidget {
  final TextBuilder _textBuilder;
  _DynamicText(this._textBuilder, {Key key}) : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return _DynamicTextState();
  }
}

class _DynamicTextState extends State<_DynamicText> {
  String _message;

  @override
  Widget build(BuildContext context) {
    return widget._textBuilder(_message);
  }

  void update([String message]) {
    if (message == _message) return; // 如果文案相同，则不刷新Text
    setState(() {
      _message = message;
    });
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_snackbar/flutter_snackbar.dart';

void main() => runApp(SnackBarApp());

class SnackBarApp extends StatelessWidget {
  GlobalKey<SnackBarWidgetState> _globalKey = GlobalKey();
  int count = 0;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        theme: ThemeData(primarySwatch: Colors.blue),
        home: Scaffold(
          appBar: AppBar(
            title: Text("SnackBar"),
            actions: <Widget>[
              InkWell(
                child: Padding(
                  child: Center(
                    child: Text("显示"),
                  ),
                  padding: EdgeInsets.only(left: 10, right: 10),
                ),
                onTap: () {
                  _globalKey.currentState.show("这是SnackBar count: ${count++}");
                },
              ),
              Padding(
                child: InkWell(
                  child: Padding(
                    child: Center(
                      child: Text("隐藏"),
                    ),
                    padding: EdgeInsets.only(left: 10, right: 10),
                  ),
                  onTap: () {
                    _globalKey.currentState.dismiss();
                  },
                ),
                padding: EdgeInsets.only(right: 10),
              )
            ],
          ),
          body: SnackBarWidget(
              // 绑定GlobalKey，用于调用显示/隐藏方法
              key: _globalKey,
              //textBuilder用于动态构建Text，用于显示变化的内容。优先级高于'text'属性
              textBuilder: (String message) {
                return Text(message ?? "",
                    style: TextStyle(color: Colors.white, fontSize: 16.0));
              },
              // 内容不变时使用text属性
              text: Text("内容不变时使用text属性"),
              // 设定背景decoration
              decoration: ShapeDecoration(
                  shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.all(Radius.circular(20))),
                  color: Colors.blue.withOpacity(0.8)),
              // 用于显示内容，默认是填充空白区域的
              content: Center(child: Text("这是内容部分"))),
        ));
  }
}
